kv-storage/src/rocksdb_kv/mod.rs

```diff
@@ -0,0 +1,300 @@
+use super::*;
+use mktemp::Temp;
+use rocksdb::{checkpoint, Options, DB};
+use std::path::PathBuf;
+
+pub struct RocksDB {
+    db: DB,
+    origin_path: PathBuf,
+    current_db_dir: Temp,
+    checkpoint_db_dir: Temp,
+}
+
+#[async_trait]
+impl KVStorage for RocksDB {
+    async fn new(path: &str) -> Result<Self, super::Error>
+    where
+        Self: Sized,
+    {
+        let origin_path = PathBuf::from(path);
+        let current_db_dir = Temp::new_dir().unwrap();
+        let checkpoint_db_dir = Temp::new_dir().unwrap();
+        {
+            let db = DB::open_default(origin_path.to_str().unwrap()).unwrap();
+            let checkpoint_db = checkpoint::Checkpoint::new(&db).unwrap();
+
+            checkpoint_db
+                .create_checkpoint(current_db_dir.to_path_buf().as_path().join("db"))
+                .unwrap();
+            checkpoint_db
+                .create_checkpoint(checkpoint_db_dir.to_path_buf().as_path().join("db"))
+                .unwrap();
+        }
+
+        Ok(RocksDB {
+            db: DB::open_default(current_db_dir.to_path_buf().as_path().join("db")).unwrap(),
+            origin_path,
+            current_db_dir,
+            checkpoint_db_dir,
+        })
+    }
+
+    async fn open(path: &str) -> Result<Self, super::Error>
+    where
+        Self: Sized,
+    {
+        let origin_path = PathBuf::from(path);
+        let current_db_dir = Temp::new_dir().unwrap();
+        let checkpoint_db_dir = Temp::new_dir().unwrap();
+        {
+            let db = DB::open_default(origin_path.to_str().unwrap()).unwrap();
+            let checkpoint_db = checkpoint::Checkpoint::new(&db).unwrap();
+
+            checkpoint_db
+                .create_checkpoint(current_db_dir.to_path_buf().as_path().join("db"))
+                .unwrap();
+            checkpoint_db
+                .create_checkpoint(checkpoint_db_dir.to_path_buf().as_path().join("db"))
+                .unwrap();
+        }
+
+        Ok(RocksDB {
+            db: DB::open_default(current_db_dir.to_path_buf().as_path().join("db")).unwrap(),
+            origin_path,
+            current_db_dir,
+            checkpoint_db_dir,
+        })
+    }
+
+    async fn commit_checkpoint(&mut self) -> Result<(), super::Error> {
+        let new_checkpoint_db_dir = Temp::new_dir().unwrap();
+        let checkpoint_db = checkpoint::Checkpoint::new(&self.db).unwrap();
+
+        DB::destroy(&Options::default(), self.origin_path.to_str().unwrap()).unwrap();
+        checkpoint_db
+            .create_checkpoint(self.origin_path.to_str().unwrap())
+            .unwrap();
+        checkpoint_db
+            .create_checkpoint(new_checkpoint_db_dir.to_path_buf().as_path().join("db"))
+            .unwrap();
```

Comment:
Don't unwrap, but handle it and propagate using `?`

Reply:
fix